test/dynamo/test_autograd_function.py

```diff
@@ -0,0 +1,255 @@
+        self.fn = CustomFunc3.apply
+
+    def forward(self, foo):
+        return self.fn(foo)
+
+
+class LinearFunction(torch.autograd.Function):
+    # Note that forward, setup_context, and backward are @staticmethods
+    @staticmethod
+    def forward(input, weight, bias):
+        output = input.mm(weight.t())
+        if bias is not None:
+            output += bias.unsqueeze(0).expand_as(output)
+        return output
+
+    @staticmethod
+    # inputs is a Tuple of all of the inputs passed to forward.
+    # output is the output of the forward().
+    def setup_context(ctx, inputs, output):
+        input, weight, bias = inputs
+        ctx.save_for_backward(input, weight, bias)
+
+    # This function has only a single output, so it gets only one gradient
+    @staticmethod
+    def backward(ctx, grad_output):
+        input, weight, bias = ctx.saved_tensors
+        grad_input = grad_weight = grad_bias = None
+        if ctx.needs_input_grad[0]:
+            grad_input = grad_output.mm(weight)
+        if ctx.needs_input_grad[1]:
+            grad_weight = grad_output.t().mm(input)
+        if bias is not None and ctx.needs_input_grad[2]:
+            grad_bias = grad_output.sum(0)
+
+        return grad_input, grad_weight, grad_bias
+
+
+class ModuleLinear(torch.nn.Module):
+    def forward(self, input, weight, bias=None):
+        return LinearFunction.apply(input, weight, bias)
+
+
+class MaterializingGradFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, x):
+        ctx.set_materialize_grads(False)
+        return x.clone(), x.clone()
+
+    @staticmethod
+    def backward(ctx, grad_out1, grad_out2):
+        return grad_out1, grad_out2
+
+
+class MaterializingGradModule(torch.nn.Module):
+    def forward(self, x):
+        return MaterializingGradFunction.apply(x)
+
+
+class CustomFuncBwdPrintGraphBreak(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, foo):
+        return torch.add(foo, foo)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        print("graph break!")
+        return grad_output
+
+
+class CustomFuncBwdPrintModule(torch.nn.Module):
+    def forward(self, x):
+        return CustomFuncBwdPrintGraphBreak.apply(x)
+
+
+class CustomFuncStrideBwd(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, foo):
+        return torch.add(foo, foo)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        return grad_output.stride()
+
+
+class CustomFuncStrideModule(torch.nn.Module):
+    def forward(self, x):
+        return CustomFuncStrideBwd.apply(x)
+
+
+class CustomFuncSaveForBwd(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, foo):
+        result = foo + foo
+        result = result + foo
+        ctx.save_for_backward(result)
+        return result
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        (result,) = ctx.saved_tensors
+        return grad_output * math.sqrt(result.numel())
+
```

Comment:
nit: it's a bit difficult to scroll up and down when reading the tests (because the tests references the modules), but it's not the worst